glean-core/src/debug.rs

```diff
@@ -18,78 +18,107 @@
+}
+
+impl std::fmt::Debug for DebugOptions {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+        fmt.debug_struct("DebugOptions")
+            .field("log_pings", &self.log_pings.get())
+            .field("debug_view_tag", &self.debug_view_tag.get())
+            .field("source_tags", &self.source_tags.get())
+            .finish()
+    }
 }
 
 impl DebugOptions {
     pub fn new() -> Self {
         Self {
-            log_pings: DebugOption::new(GLEAN_LOG_PINGS, None),
-            debug_view_tag: DebugOption::new(GLEAN_DEBUG_VIEW_TAG, Some(validate_debug_view_tag)),
+            log_pings: DebugOption::new(GLEAN_LOG_PINGS, get_bool_from_str, None),
+            debug_view_tag: DebugOption::new(GLEAN_DEBUG_VIEW_TAG, Some, Some(validate_tag)),
+            source_tags: DebugOption::new(
+                GLEAN_SOURCE_TAGS,
+                tokenize_string,
+                Some(validate_source_tags),
+            ),
         }
     }
 }
 
 /// A representation of a debug option,
 /// where the value can be set programmatically or come from an environment variable.
 #[derive(Debug)]
-pub struct DebugOption<T, F = fn(T) -> Option<T>> {
+pub struct DebugOption<T, E = fn(String) -> Option<T>, F = fn(&T) -> bool> {
     /// The name of the environment variable related to this debug option.
     env: String,
     /// The actual value of this option.
     value: Option<T>,
-    /// Optional function to validade the value parsed from the environment
+    /// Function to extract the data of type `T` from a `String`, used when
+    /// extracting data from the environment.
+    extraction: E,
+    /// Optional function to validate the value parsed from the environment
     /// or passed to the `set` function.
     validation: Option<F>,
 }
 
-impl<T, F> DebugOption<T, F>
+impl<T, E, F> DebugOption<T, E, F>
 where
-    T: std::str::FromStr,
-    F: Fn(T) -> Option<T>,
+    T: Clone,
+    E: Fn(String) -> Option<T>,
+    F: Fn(&T) -> bool,
 {
     /// Create a new debug option,
     /// tries to get the initial value of the option from the environment.
-    pub fn new(env: &str, validation: Option<F>) -> Self {
+    pub fn new(env: &str, extraction: E, validation: Option<F>) -> Self {
         let mut option = Self {
             env: env.into(),
             value: None,
+            extraction,
             validation,
         };
 
         option.set_from_env();
         option
     }
 
-    fn validate(&self, value: T) -> Option<T> {
+    fn validate(&self, value: &T) -> bool {
         if let Some(f) = self.validation.as_ref() {
             f(value)
         } else {
-            Some(value)
+            true
         }
     }
 
     fn set_from_env(&mut self) {
+        let extract = &self.extraction;
         match env::var(&self.env) {
-            Ok(env_value) => match T::from_str(&env_value) {
-                Ok(v) => {
+            Ok(env_value) => match extract(env_value.clone()) {
```

Comment:
That `clone` seems unnecessary (we do own `env_value` here), but it's also not that big of a deal. This happens once and only for debugging.

Reply:
You're right :) Moreover, if we change things to use a `&` here, the `Some` trick instead of the `identity` function won't work anymore. I'd rather leave it as it is and eventually change it if it becomes a problem.